In [1]:
import numpy as np


Cross entropy

In [6]:
def cross_entropy(predictions, targets, epsilon=1e-12):
    """
    Computes cross entropy between targets (encoded as one-hot vectors)
    and predictions.
    Input: predictions (N, k) ndarray
           targets (N, k) ndarray
    Returns: scalar
    """
    predictions = np.clip(predictions, epsilon, 1. - epsilon)
    N = predictions.shape[0]
    ce = -np.sum(targets*np.log(predictions+1e-9))/N
    return ce


here a simulated prediction of GP

you can see that prediction are lower, not all are 0.9 and above

In [9]:
predictions = np.array([
                        [0.25,0.25,0.25,0.60],
                        [0.01,0.01,0.01,0.60],
                        [0.25,0.25,0.25,0.90],
                        [0.01,0.01,0.01,0.88],
                        [0.25, 0.25, 0.25, 0.70],
                        [0.01, 0.01, 0.01, 0.66],
                        [0.01, 0.01, 0.01, 0.55],
                        [0.01, 0.5, 0.10, 0.66],
                        [0.01, 0.01, 0.01, 0.67],
                        [0.01, 0.01, 0.01, 0.67],
                        [0.25, 0.25, 0.25, 0.90],
                        [0.01, 0.01, 0.01, 0.88],
                        [0.25, 0.25, 0.25, 0.70],
                        [0.01, 0.01, 0.01, 0.66],
                        [0.01, 0.01, 0.01, 0.55],
                        [0.01, 0.5, 0.10, 0.66],
                        [0.01, 0.01, 0.01, 0.67],
                        [0.01, 0.01, 0.01, 0.67],
                        [0.25, 0.25, 0.25, 0.90],
                        [0.01, 0.01, 0.01, 0.88],
                        [0.25, 0.25, 0.25, 0.70],
                        [0.01, 0.01, 0.01, 0.66],
                        [0.01, 0.01, 0.01, 0.55],
                        [0.01, 0.5, 0.10, 0.66],
                        [0.01, 0.01, 0.01, 0.67],
                        [0.01, 0.01, 0.01, 0.67]
])
targets = np.array([
                        [0,0,0,1],
                        [0,0,0,1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
])

simulated overfitted network, as you can see, all predictions are very high

And the main point is, that you have 2 false classified examples

In [10]:
predictions2 = np.array([
                        [0.25,0.25,0.25,0.99],
                        [0.01,0.01,0.01,0.99],
                        [0.25,0.25,0.25,0.99],
                        [0.01,0.01,0.01,0.99],
                        [0.25, 0.25, 0.25, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.5, 0.90, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.25, 0.25, 0.25, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.25, 0.25, 0.25, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.5, 0.90, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.25, 0.25, 0.25, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.25, 0.25, 0.25, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.5, 0.90, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
                        [0.01, 0.01, 0.01, 0.99],
])

targets2 = np.array([
                        [0,0,1,0],
                        [0,0,1,0],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
                        [0, 0, 0, 1],
])


In [11]:
from sklearn.metrics import accuracy_score

print(predictions.argmax(1))
print(targets.argmax(1))

print(predictions2.argmax(1))
print(targets2.argmax(1))

x_gp = cross_entropy(predictions, targets)
print("gp simulation CE:", x_gp)
print("gp simulation Acc:", accuracy_score(predictions.argmax(1), targets.argmax(1)))

x = cross_entropy(predictions2, targets2)
print("overfitting simulation CE:", x)
print("overfitting simulation Acc:", accuracy_score(predictions2.argmax(1), targets2.argmax(1)))


[3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3]
[3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3]
[3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3]
[2 2 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3 3]
gp simulation CE: 0.36464333725418346
gp simulation Acc: 1.0
overfitting simulation CE: 0.23971817228267683
overfitting simulation Acc: 0.9230769230769231


as you can see Acc is higher but the CE is lower

gp simulation CE: 0.36464333725418346

gp simulation Acc: 1.0

overfitting simulation CE: 0.23971817228267683

overfitting simulation Acc: 0.9230769230769231
